### DISCUSSION TITLE
Guide : Upscaler database + Comparison

### DISCUSSION THREAD

**Discussion by glucauze** [glucauze/sd-webui-faceswaplab#29] [Guide/Doc]
Here are two sources of information that I find really interesting on upscaling:
+ [An upscaler database](https://upscale.wiki/wiki/Model_Database)
+ [A really well written guide and some comparisons](https://phhofm.github.io/upscale/favorites.html). 

Getting good results is a process of testing and trying different methods.

The [003_realSR_BSRGAN_DFOWMFC_s64w8_SwinIR-L_x4_GAN](https://github.com/JingyunLiang/SwinIR/releases/download/v0.0/003_realSR_BSRGAN_DFOWMFC_s64w8_SwinIR-L_x4_GAN.pth) model seems to yield good results in a reasonable amount of time for post-processing. Combined with the GPU options in version 1.2.1 (not yet released), the processing becomes really fast.

**Comment by glucauze** [comment]
From my testing  in postprocessing
```
Prompt: portrait of a man
Steps: 20 | Seed: 2461472201 | Sampler: UniPC | CFG scale: 6 | Size: 512x512 | Parser: Full parser | Model: dreamshaper_8 | Model hash: 879db523c3 | Version: 6427316 | Pipeline: Original | Operations: txt2img
```

+ Orig Image (before swap)

![orig](https://github.com/glucauze/sd-webui-faceswaplab/assets/137925069/ab4a6304-f438-41c6-9e4e-77ca1907ca1d)

+ No post-processing or global-postprocessing (reference) : 

![nothing](https://github.com/glucauze/sd-webui-faceswaplab/assets/137925069/44951f55-b372-463c-bca1-6f8951a494a5)

+ No post-processing + CodeFormer in Global postprocessing : 

![nopost_codeformer](https://github.com/glucauze/sd-webui-faceswaplab/assets/137925069/8a91dc4f-5ef8-4ac9-a1e9-777971cb3c84)

**Post-processing only (no global post-processing), all done with sharpen, color correction and buggy improved mask (affected by color correction in these examples, will be fixed) :**

+ TGHQFace8x_500k + GFPGAN (very ugly with codeformer)

![TGHQFace8x_500k](https://github.com/glucauze/sd-webui-faceswaplab/assets/137925069/4bcda763-08f5-4419-9b9a-bc89533100bd)

+ 4x_RealisticRescaler_100000_G + GFPGAN : 

![4x_RealisticRescaler_100000_G](https://github.com/glucauze/sd-webui-faceswaplab/assets/137925069/82e54672-d1b4-4305-92ab-69c8b2bd9823)

+ SwinIR_4x + Codeformer (slower)

![swinir](https://github.com/glucauze/sd-webui-faceswaplab/assets/137925069/2bb4ae97-2875-4b91-b308-45e16fe37bab)

+ 003_realSR_BSRGAN_DFOWMFC_s64w8_SwinIR-L_x4_GAN + Codeformer (slower)

![003_realSR_BSRGAN_DFOWMFC_s64w8_SwinIR-L_x4_GAN](https://github.com/glucauze/sd-webui-faceswaplab/assets/137925069/52d291ef-81c1-4655-816d-d65ad32e0dee)

+ LDSR + CodeFormer (very slow)

![ldsr](https://github.com/glucauze/sd-webui-faceswaplab/assets/137925069/a8334b54-69eb-4aa4-986b-cc8e7c7d6730)

Post-Processing example settings : 

![settings](https://github.com/glucauze/sd-webui-faceswaplab/assets/137925069/1b92981b-2f2d-406b-ba03-c6d5b2bb37a5)

**Comment by glucauze** [comment]
Improved mask comparison (working version)

Without : 

![no_improved_mask](https://github.com/glucauze/sd-webui-faceswaplab/assets/137925069/43a8278c-885a-4663-b152-9327b68b2f84)


With :

![improved_mask](https://github.com/glucauze/sd-webui-faceswaplab/assets/137925069/054f1e25-99e3-4028-ba85-d4c74245d46d)

**Comment by ColinFrake** [comment]
I downloaded [003_realSR_BSRGAN_DFOWMFC_s64w8_SwinIR-L_x4_GAN](https://github.com/JingyunLiang/SwinIR/releases/download/v0.0/003_realSR_BSRGAN_DFOWMFC_s64w8_SwinIR-L_x4_GAN.pth) model, where should I put this file? Thank you very much!

**Reply by glucauze** [reply]
It depends on the model but most of them go to <webui>/models/ESRGAN (whey they are based on ESRGAN which is the case here)

**Comment by agarzon** [comment]
I tend to use 4x-UltraSharp for everything, am I doing wrong? I'm 100% into realistic portraits, 0% anime, cartoon or 3d renderings.

**Reply by glucauze** [reply]
No it only depends on the subject and your personal preferences. You can get good results with this upscaler.

In the code I've done (**in post-processing & advanced mask option**), the upscaling of the face is done with a factor of 4 before face restore and merging the swapped face with the original image. We go from a 128x128 image with little definition to 512x512. Some upscalers, which generally work very well, may not be ideal for this application. That's why you need to test different solutions and choose the one that seems best for you.

For the global post processing, it's up to your liking. It will upscale the whole image.

**Reply by agarzon** [reply]
Thanks.

My personal chooses for realistic upscalers are so far 4x-UltraSharp and 4x_NMKD-Siax_200k (yes they are slow), if anyone knows something better please let us know 😇

I have a question for you: I was looking in the config.json and the ui-config.json but I did not find a way to set the **Upscaler scale**, let's say I want all my productions to be x1.5 permanently 

So far I did not find where to set that as default value, I have to change it each time manually in the Global Post-Processing tab

Do you know how? thanks a bunch

![image](https://github.com/glucauze/sd-webui-faceswaplab/assets/339828/fdfd1a9e-83d6-4232-839e-46a2b5f16488)

**Reply by glucauze** [reply]
Yes it's normal. I'll try to change the option for the scale

**Comment by Creepybits** [comment]
I keep getting the following error after downloading the upscaler from your link.

_Failed loading SwinIR model C:\AI Stuff\A1111\stable-diffusion-webui\models\SwinIR\003_realSR_BSRGAN_DFO_s64w8_SwinIR-M_x2_GAN.pth: 'params_ema'_

It shows up in the list without any issues though.

![image](https://github.com/glucauze/sd-webui-faceswaplab/assets/60452756/63951f09-a0d2-4e23-9a3d-3101631ec7f2)

**Comment by ugintl** [comment]
Is it still working in 2025?

I am using python 3.10 and comfyui.

Main images I am working on are oil painting, impressionist, watercolor and realistic.

I generated them using kolors model and now I want to upscale.

**Reply by agarzon** [reply]
This project was abandoned 2 years ago. Look for something called Reactor in Comfyui.
